src/Database/LSMTree/Internal/Unsliced.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE DeriveFunctor     #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RoleAnnotations   #-}
+
+module Database.LSMTree.Internal.Unsliced (
+    -- * Unsliced raw bytes
+    Unsliced
+    -- * Unsliced keys
+  , makeUnslicedKey
+  , unsafeMakeUnslicedKey
+  , fromUnslicedKey
+  , unsafeNoAssertMakeUnslicedKey
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Primitive.ByteArray
+import qualified Data.Vector.Primitive as PV
+import           Database.LSMTree.Internal.RawBytes (RawBytes (..))
+import qualified Database.LSMTree.Internal.RawBytes as RB
+import           Database.LSMTree.Internal.Serialise (SerialisedKey (..))
+import           Database.LSMTree.Internal.Vector (mkPrimVector,
+                     noRetainedExtraMemory)
+
+-- | Unsliced string of bytes
+type role Unsliced nominal
+newtype Unsliced a =
+    Unsliced ByteArray
+
+getByteArray :: RawBytes -> ByteArray
+getByteArray (RawBytes (PV.Vector _ _ ba)) = ba
+
+precondition :: RawBytes -> Bool
+precondition (RawBytes pvec) = noRetainedExtraMemory pvec
+
+makeUnsliced :: RawBytes -> Unsliced RawBytes
+makeUnsliced bytes
+    | precondition bytes = Unsliced (getByteArray bytes)
+    | otherwise     = Unsliced (getByteArray $ RB.copy bytes)
+
+unsafeMakeUnsliced :: RawBytes -> Unsliced RawBytes
+unsafeMakeUnsliced bytes = assert (precondition bytes) (Unsliced (getByteArray bytes))
+
+unsafeNoAssertMakeUnsliced :: RawBytes -> Unsliced RawBytes
+unsafeNoAssertMakeUnsliced bytes = Unsliced (getByteArray bytes)
+
+fromUnsliced :: Unsliced RawBytes -> RawBytes
+fromUnsliced (Unsliced ba) = RawBytes (mkPrimVector 0 (sizeofByteArray ba) ba)
+
+{-------------------------------------------------------------------------------
+  Unsliced keys
+-------------------------------------------------------------------------------}
+
+from :: Unsliced RawBytes -> Unsliced SerialisedKey
+from (Unsliced ba) = Unsliced ba
+
+to :: Unsliced SerialisedKey -> Unsliced RawBytes
+to (Unsliced ba) = Unsliced ba
```

Comment:
The other functions all have a `Key` prefix, I think these ones need it, too, to distinguish them from potential future `Unsliced RawBytes -> Unsliced SerialisedValue` etc.

More generally, I'm not sure how useful it is to go via `Unsliced RawBytes` at all, if we only use `Unsliced SerialisedKey` and need to copy the whole API for each new supported type anyway. But I don't mind going with it for now, new use cases might still show up.

Reply:
> The other functions all have a `Key` prefix, I think these ones need it, too, to distinguish them from potential future `Unsliced RawBytes -> Unsliced SerialisedValue` etc.

Good point. I'll see if I can include it as an aside in a future PR

> More generally, I'm not sure how useful it is to go via `Unsliced RawBytes` at all, if we only use `Unsliced SerialisedKey` and need to copy the whole API for each new supported type anyway. But I don't mind going with it for now, new use cases might still show up.

I'd argue that having the checks/assertion logic in one place (for the `Unsliced Rawbytes`) is useful, so we don't have to repeat it for all three `Serialised*` newtypes. The API is duplicated but it's only wrapping/unwrapping things

Truthfully, we might not even need this level of generality. Maybe it's only `UnslicedSerialisedKey` that we need, or maybe we don't need a newtype at all, though I'd argue enforcing the slicing checks/assertions is helpful